Tests/GRPCTests/GRPCKeepaliveTests.swift

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2020, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+@testable import GRPC
+import NIO
+import EchoModel
+import EchoImplementation
+import XCTest
+
+class GRPCKeepaliveTests: GRPCTestCase {
+  func testClientKeepaliveTimeout() {
+    XCTAssertNoThrow(try self.doTestKeepalive(
+      clientKeepalive: ClientConnectionKeepalive(time: .seconds(3), timeout: .seconds(1)),
+      expectedFinalChange: .shutdown)
+    )
+  }
+
+  func doTestKeepalive(clientKeepalive: ClientConnectionKeepalive = ClientConnectionKeepalive(),
+                       serverKeepalive: ServerConnectionKeepalive = ServerConnectionKeepalive(),
+                       expectedFinalChange: ConnectivityState) throws {
+    let group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+    defer {
+      XCTAssertNoThrow(try group.syncShutdownGracefully())
+    }
+
+    // Setup a server.
+    let server = try Server.insecure(group: group)
+      .withServiceProviders([DelayedEchoProvider(delay: 60)])
+      .withKeepalive(serverKeepalive)
+      .bind(host: "localhost", port: 0)
+      .wait()
+    defer {
+      XCTAssertNoThrow(try server.close().wait())
+    }
+
+    // Setup a state change recorder for the client.
+    let stateRecorder = RecordingConnectivityDelegate()
+    stateRecorder.expectChanges(3) { changes in
+      XCTAssertEqual(changes, [
+        Change(from: .idle, to: .connecting),
+        Change(from: .connecting, to: .ready),
+        Change(from: .ready, to: .shutdown)
```

Comment:
Hmm, `.shutdown` shouldn't be the final state. It should be `.idle`. I.e. "we don't have a connection now but we could have one in the future" whereas shutdown is "we don't have a connection now and we'll never have one in the future".